kafka/channel/pkg/dispatcher/dispatcher.go

```diff
@@ -303,33 +320,80 @@ func (d *KafkaDispatcher) getChannelReferenceFromHost(host string) (eventingchan
 	return cr, nil
 }
 
-func fromKafkaMessage(kafkaMessage *sarama.ConsumerMessage) *contribchannels.Message {
-	headers := make(map[string]string, len(kafkaMessage.Headers))
+func fromKafkaMessage(ctx context.Context, kafkaMessage *sarama.ConsumerMessage) *cloudevents.Event {
+	var event = &cloudevents.Event{}
 	for _, header := range kafkaMessage.Headers {
-		headers[string(header.Key)] = string(header.Value)
-	}
-	message := contribchannels.Message{
-		Headers: headers,
-		Payload: kafkaMessage.Value,
+		h := string(header.Key)
+		v := string(header.Value)
+		logging.FromContext(ctx).Debugf("key: %s, value %s", h, v)
+		switch h {
+		case "ce_datacontenttype":
+			event.SetDataContentType(v)
+		case "ce_specversion":
+			event.SetSpecVersion(v)
+		case "ce_type":
+			event.SetType(v)
+		case "ce_source":
+			event.SetSource(v)
+		case "ce_id":
+			event.SetID(v)
+		case "ce_time":
+			t, _ := time.Parse(time.RFC3339, v)
+			event.SetTime(t)
+		case "ce_subject":
+			event.SetSubject(v)
+		case "ce_dataschema":
+			event.SetDataSchema(v)
+		default:
+			// Extensions
+			event.SetExtension(strings.ToLower(h), strings.ToLower(v))
+		}
 	}
-	return &message
+	return event
 }
 
-func toKafkaMessage(channel eventingchannels.ChannelReference, message *contribchannels.Message, topicFunc TopicFunc) *sarama.ProducerMessage {
-	kafkaMessage := sarama.ProducerMessage{
-		Topic:   topicFunc(utils.KafkaChannelSeparator, channel.Namespace, channel.Name),
-		Value:   sarama.ByteEncoder(message.Payload),
-		Headers: make([]sarama.RecordHeader, len(message.Headers)),
+func toKafkaMessage(ctx context.Context, channel eventingchannels.ChannelReference, event cloudevents.Event, topicFunc TopicFunc) *sarama.ProducerMessage {
+	data, err := event.DataBytes()
+	if err != nil {
+		return &sarama.ProducerMessage{} //this should probably be something else indicating an error
 	}
-	i := 0
-	for h, v := range message.Headers {
-		kafkaMessage.Headers[i].Key = []byte(h)
-		kafkaMessage.Headers[i].Value = []byte(v)
-		i++
+
+	kafkaMessage := sarama.ProducerMessage{
+		Topic: topicFunc(utils.KafkaChannelSeparator, channel.Namespace, channel.Name),
+		Value: sarama.ByteEncoder(data),
 	}
+	kafkaMessage = attachKafkaHeaders(kafkaMessage, event)
 	return &kafkaMessage
 }
 
+func attachKafkaHeaders(message sarama.ProducerMessage, event cloudevents.Event) sarama.ProducerMessage {
+	addHeader(&message, "ce_specversion", event.SpecVersion())
+	addHeader(&message, "ce_type", event.Type())
+	addHeader(&message, "ce_source", event.Source())
+	addHeader(&message, "ce_id", event.ID())
+	addHeader(&message, "ce_time", event.Time().Format(time.RFC3339))
+	if event.DataContentType() != "" {
+		addHeader(&message, "ce_datacontenttype", event.DataContentType())
+	}
+	if event.Subject() != "" {
+		addHeader(&message, "ce_subject", event.Subject())
```

Comment:
there is no subject mentioned here:
https://github.com/cloudevents/spec/blob/v1.0/kafka-protocol-binding.md#325-example

@lberk can u follow up w/ an issue on the CE spec, for kafka? - looks like their 1.0 lacks some attrs for Kafka, like `subject`